File: demo/nodejs/gravity.js
/*
    gravity.js  -  by Don Cross  -  2021-07-19

    Example Node.js program for Astronomy Engine:
    https://github.com/cosinekitty/astronomy
*/

const UsageText = `
    USAGE:

    node gravity.js latitude height

    Calculates the gravitational acceleration experienced
    by an observer on the surface of the Earth at the specified
    latitude (degrees north of the equator) and height
    (meters of above sea level).
    The output is the gravitational acceleration in m/s^2.
`;

const Astronomy = require('./astronomy.js');

function ParseNumber(name, text, minValue, maxValue) {
    const x = Number(text);
    if (!Number.isFinite(x) || (x < minValue) || (x > maxValue)) {
        console.error(`ERROR: Not a valid numeric value for ${name}: "${text}".`);
        console.error(`Must be in the range ${minValue} .. ${maxValue}.`);
        process.exit(1);
    }
    return x;
}

function Format(value, width, precision) {
    let s = value.toFixed(precision);
    while (s.length < width) {
        s = ' ' + s;
    }
    return s;
}

function Demo() {
    if (process.argv.length !== 4) {
        console.log(UsageText);
        process.exit(1);
    } else {
        const latitude = ParseNumber('latitude', process.argv[2], -90, +90);
        const height = ParseNumber('height', process.argv[3], 0, 100000);
        const gravity = Astronomy.ObserverGravity(latitude, height);
        console.log(`latitude = ${Format(latitude, 8, 4)},  height = ${Format(height, 6, 0)},  gravity = ${Format(gravity, 8, 6)}`);
        process.exit(0);
    }
}

Demo();

File: demo/python/gravity.py
#!/usr/bin/env python3
import sys
from astronomy import ObserverGravity

UsageText = r'''

    USAGE:

    gravity.py latitude height

    Calculates the gravitational acceleration experienced
    by an observer on the surface of the Earth at the specified
    latitude (degrees north of the equator) and height
    (meters above sea level).
    The output is the gravitational acceleration in m/s^2.

'''

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print(UsageText)
        sys.exit(1)

    latitude = float(sys.argv[1])
    if latitude < -90.0 or latitude > +90.0:
        print("ERROR: Invalid latitude '{}'. Must be a number between -90 and +90.".format(sys.argv[1]))
        sys.exit(1)

    height = float(sys.argv[2])
    MAX_HEIGHT_METERS = 100000.0
    if height < 0.0 or height > MAX_HEIGHT_METERS:
        print("ERROR: Invalid height '{}'. Must be a number between 0 and {}.".format(sys.argv[1], MAX_HEIGHT_METERS))
        sys.exit(1)

    gravity = ObserverGravity(latitude, height)
    print("latitude = {:8.4f},  height = {:6.0f},  gravity = {:8.6f}".format(latitude, height, gravity))
    sys.exit(0)
